### PR TITLE
Fail more gracefully when option collections cleared

### PR DESCRIPTION
**Pull request by dougbu**
- #4690
- move `ModelBindingMessageProvider` init from `DefaultBindingMetadataProvider` to `DefaultModelMetadata`
 - in addition to avoiding error cases, this removes some boilerplate
- add specific errors to `BodyModelBinderProvider`, `CompositeViewEngine`,
  `DefaultRazorViewEngineFileProviderAccessor`, `ModelBinderFactory`, and `ObjectResultExecutor`

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreMvcOptionsSetup.cs [hunk 1]
```diff
@@ -79,7 +67,7 @@ public void Configure(MvcOptions options)
             // by altering the collection of providers.
             options.ModelMetadataDetailsProviders.Add(new ExcludeBindingMetadataProvider(typeof(Type)));
 
-            options.ModelMetadataDetailsProviders.Add(new DefaultBindingMetadataProvider(messageProvider));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreMvcOptionsSetup.cs]: This ends up being really good cleanup :+1:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/EmptyModelMetadataProvider.cs [hunk 2]
```diff
@@ -1,52 +1,25 @@
...
-                new MessageOnlyBindingProvider()
-            }))
+            : base(
+                  new DefaultCompositeMetadataDetailsProvider(new List<IMetadataDetailsProvider>()),
+                  new OptionsAccessor())
         {
         }
 
-        private class MessageOnlyBindingProvider : IBindingMetadataProvider
+        private class OptionsAccessor : IOptions<MvcOptions>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/EmptyModelMetadataProvider.cs]: We already have `TestOptionsManager<>` that does this

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/EmptyModelMetadataProvider.cs [hunk 3]
```diff
@@ -1,52 +1,25 @@
...
-                {
-                    MissingBindRequiredValueAccessor = Resources.FormatModelBinding_MissingBindRequiredMember,
-                    MissingKeyOrValueAccessor = Resources.FormatKeyValuePair_BothKeyAndValueMustBePresent,
-                    ValueMustNotBeNullAccessor = Resources.FormatModelBinding_NullValueNotValid,
-                    AttemptedValueIsInvalidAccessor = Resources.FormatModelState_AttemptedValueIsInvalid,
-                    UnknownValueIsInvalidAccessor = Resources.FormatModelState_UnknownValueIsInvalid,
-                    ValueIsInvalidAccessor = Resources.FormatHtmlGeneration_ValueIsInvalid,
-                    ValueMustBeANumberAccessor = Resources.FormatHtmlGeneration_ValueMustBeNumber,
-                };
-            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/EmptyModelMetadataProvider.cs]: yay, loving this cleanup

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 4]
```diff
@@ -32,10 +33,27 @@ public class DefaultModelMetadata : ModelMetadata
         /// <param name="provider">The <see cref="IModelMetadataProvider"/>.</param>
         /// <param name="detailsProvider">The <see cref="ICompositeMetadataDetailsProvider"/>.</param>
         /// <param name="details">The <see cref="DefaultMetadataDetails"/>.</param>
+        /// <remarks>Use other constructor overload. This is provided for back compatibility and a few tests.</remarks>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: I don't really feel like this is necessary or accurate. The API shape makes it clear what the tradeoff is here.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 5]
```diff
@@ -82,6 +106,12 @@ public BindingMetadata BindingMetadata
                 if (_details.BindingMetadata == null)
                 {
                     var context = new BindingMetadataProviderContext(Identity, _details.ModelAttributes);
+
+                    // Provide a unique ModelBindingMessageProvider instance based on the one passed to the constructor
+                    // (from MvcOptions).
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: This comment explains what, not why. Omit it or explain why we need to do this.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs [hunk 6]
```diff
@@ -17,17 +18,42 @@ public class DefaultModelMetadataProvider : IModelMetadataProvider
...
+            ModelBindingMessageProvider messageProvider)
+        {
+            if (detailsProvider == null)
+            {
+                throw new ArgumentNullException(nameof(detailsProvider));
+            }
+
+            DetailsProvider = detailsProvider;
+            _cacheEntryFactory = CreateCacheEntry;
+            _messageProvider = messageProvider;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: This needs to be exposed in protected property so this class can be extended

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs [hunk 7]
```diff
@@ -46,12 +46,20 @@ public IModelBinder CreateBinder(ModelBinderFactoryContext context)
                 throw new ArgumentNullException(nameof(context));
             }
 
+            if (_providers.Length == 0)
+            {
+                throw new InvalidOperationException(Resources.FormatModelBinderProvidersAreRequired(
+                    typeof(MvcOptions).FullName,
+                    nameof(MvcOptions.ModelBinderProviders),
+                    typeof(IModelBinderProvider).FullName));
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs]: Why is this necessary? without this change we still report a good error message right?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs [hunk 8]
```diff
@@ -18,7 +19,15 @@ public class DefaultRazorViewEngineFileProviderAccessor : IRazorViewEngineFilePr
...
         {
             var fileProviders = optionsAccessor.Value.FileProviders;
-            if (fileProviders.Count == 1)
+            if (fileProviders.Count == 0)
+            {
+                var message = Resources.FormatFileProvidersAreRequired(
+                    typeof(RazorViewEngineOptions).FullName,
+                    nameof(RazorViewEngineOptions.FileProviders),
+                    typeof(IFileProvider).FullName);
+                throw new ArgumentException(message, nameof(optionsAccessor));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs]: I don't understand why this is an error. If the list of file providers is empty you'll still have a reasonable and well defined experience

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs]: The fact that this is in the constructor of something from DI makes me concerned.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs [hunk 9]
```diff
@@ -77,6 +85,14 @@ public ViewEngineResult GetView(string executingFilePath, string viewPath, bool
                 throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(viewPath));
             }
 
+            if (ViewEngines.Count == 0)
+            {
+                throw new InvalidOperationException(Resources.FormatViewEnginesAreRequired(
+                    typeof(MvcViewOptions).FullName,
+                    nameof(MvcViewOptions.ViewEngines),
+                    typeof(IViewEngine).FullName));
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs]: In contrast to the file provider part - I think this is right. If you get here it's because you tried to render a view, which will never succeed for your app.

**Comment by rynowak** [comment]
Most of this seems pretty solid :watch:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/EmptyModelMetadataProvider.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/EmptyModelMetadataProvider.cs]: :see_no_evil: this is product code.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: You mean something like `Provide a unique ModelBindingMessageProvider instance so providers can customize it per-type.`?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: Well, it's accurate but I can remove.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs (hunk 6 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: :question: There is nothing `protected` or `virtual` in this class.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs (hunk 7 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs]: The error w/o this says nothing about options. It only reports an inability to create a specific binder.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs]: 1. If this list of file providers is empty, executing any `ViewResult` throws an `InvalidOperationException` stating it couldn't find the specified view in one or more searched locations. I.e. the behaviour is identical to the case where the file is actually missing and the error doesn't point users toward the cause.
2. This provider is used only in code directly related to finding and compiling views. And the `Exception` bubbles up correctly. Agree that we previously had issues with less-than-useful (i.e. totally obfuscating) errors due to exceptions in service constructors but that's no longer the case.
3. Can't check for an empty provider collection in the consuming code due to back-compat: Must leave `IRazorViewEngineFileProviderAccessor` unchanged and down-casting to check for a new `DefaultRazorViewEngineFileProviderAccessor` property (e.g. `public int Count { get; }`) is just ugly.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs (hunk 9 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs]: Disagree w/ the contrast you're making. Note the `IRazorViewEngineFileProviderAccessor` name and how it's used.

**Comment by dougbu** [comment]
:up::date: and responded to comments

**Comment by dougbu** [comment]
(assigning to @rynowak since he seems to have volunteered :smiley_cat:)

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: https://github.com/aspnet/Mvc/pull/4928/files#diff-c8fa65ecc8077ca10cd5f923c0082de2R62 among others. There's lots of virtual stuff

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs]: :ok: I don't think that this is harmful

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorViewEngineFileProviderAccessor.cs]: :ok: I think I'm ok with the code as-is. If someone really needs this we could make the getter throw, or just remove the error.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs]: Constructors are very different than a method like `FindView`. Since we use DI + ctor injection that means that constructors should rarely throw.

**Comment by rynowak** [comment]
:shipit: after resolving https://github.com/aspnet/Mvc/pull/4928/files#r69019993

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs (hunk 6 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: :ok: I totally misread where your comment pointed.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs (hunk 9 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs]: Your comment here seems to conflict w/ the one [there](https://github.com/aspnet/Mvc/pull/4928/files#r69055323). Let's resolve this tomorrow morning.

**Comment by dougbu** [comment]
:up::date:

**Comment by dougbu** [comment]
:up::date: @rynowak I've resolved your DI-induced headache.

**Comment by rynowak** [comment]
:shipit:

**Comment by dougbu** [comment]
42cea41
